File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/retrieval/specs/GivenATypicalLibrary/WhenCollectingTheStoredFiles.java
package com.lasthopesoftware.bluewater.client.stored.library.items.files.retrieval.specs.GivenATypicalLibrary;

import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFileEntityInformation;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.retrieval.StoredFilesCollection;
import com.lasthopesoftware.bluewater.repository.InsertBuilder;
import com.lasthopesoftware.bluewater.repository.RepositoryAccessHelper;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;
import com.lasthopesoftware.specs.AndroidContext;
import org.junit.Test;
import org.robolectric.RuntimeEnvironment;

import java.util.Collection;

import static org.assertj.core.api.Assertions.assertThat;

public class WhenCollectingTheStoredFiles extends AndroidContext {
	private static Collection<StoredFile> storedFiles;

	@Override
	public void before() throws Exception {
		try(final RepositoryAccessHelper repositoryAccessHelper = new RepositoryAccessHelper(RuntimeEnvironment.application)) {
			final String insertSql = InsertBuilder
				.fromTable(StoredFileEntityInformation.tableName)
				.addColumn(StoredFileEntityInformation.serviceIdColumnName)
				.addColumn(StoredFileEntityInformation.libraryIdColumnName)
				.addColumn(StoredFileEntityInformation.isOwnerColumnName)
				.build();

			for (int i = 1; i < 10; i++) {
				repositoryAccessHelper
					.mapSql(insertSql)
					.addParameter(StoredFileEntityInformation.serviceIdColumnName, i)
					.addParameter(StoredFileEntityInformation.libraryIdColumnName, 2)
					.addParameter(StoredFileEntityInformation.isOwnerColumnName, true)
					.execute();
			}

			for (int i = 13; i < 23; i++) {
				repositoryAccessHelper
					.mapSql(insertSql)
					.addParameter(StoredFileEntityInformation.serviceIdColumnName, i)
					.addParameter(StoredFileEntityInformation.libraryIdColumnName, 5)
					.addParameter(StoredFileEntityInformation.isOwnerColumnName, true)
					.execute();
			}
		}

		final StoredFilesCollection storedFilesCollection = new StoredFilesCollection(RuntimeEnvironment.application);
		storedFiles = new FuturePromise<>(storedFilesCollection.promiseAllStoredFiles(new Library().setId(5))).get();
	}

	@Test
	public void thenTheStoredFilesAreFromTheCorrectLibrary() {
		assertThat(Stream.of(storedFiles).map(StoredFile::getLibraryId).toList()).containsOnly(5);
	}

	@Test
	public void thenTheStoredFilesAreCorrect() {
		assertThat(Stream.of(storedFiles).map(StoredFile::getServiceId).toList()).containsOnlyElementsOf(Stream.range(13, 23).toList());
	}
}

File: integrationTests/src/test/kotlin/com/lasthopesoftware/bluewater/client/stored/library/items/files/retrieval/GivenATypicalLibrary/WhenCollectingTheStoredFiles.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.files.retrieval.GivenATypicalLibrary

import androidx.test.core.app.ApplicationProvider
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.stored.library.items.files.StoredFileAccess
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFileEntityInformation
import com.lasthopesoftware.bluewater.repository.InsertBuilder
import com.lasthopesoftware.bluewater.repository.RepositoryAccessHelper
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.resources.executors.ThreadPools
import com.namehillsoftware.lazyj.Lazy
import org.assertj.core.api.Assertions
import org.junit.AfterClass
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class WhenCollectingTheStoredFiles {

	companion object {
		private var storedFiles: Lazy<Collection<StoredFile>?>? = Lazy {
            val insertSql = InsertBuilder.fromTable(StoredFileEntityInformation.tableName)
                .addColumn(StoredFileEntityInformation.serviceIdColumnName)
                .addColumn(StoredFileEntityInformation.libraryIdColumnName)
                .addColumn(StoredFileEntityInformation.isOwnerColumnName)
                .build()

            ThreadPools.promiseTableMessage<Unit, StoredFile> {
                RepositoryAccessHelper(ApplicationProvider.getApplicationContext()).use { repositoryAccessHelper ->
                    repositoryAccessHelper.beginTransaction().use {
                        for (i in 1..9) {
                            repositoryAccessHelper
                                .mapSql(insertSql)
                                .addParameter(StoredFileEntityInformation.serviceIdColumnName, i)
                                .addParameter(StoredFileEntityInformation.libraryIdColumnName, 2)
                                .addParameter(StoredFileEntityInformation.isOwnerColumnName, true)
                                .execute()
                        }
                        for (i in 13..22) {
                            repositoryAccessHelper
                                .mapSql(insertSql)
                                .addParameter(StoredFileEntityInformation.serviceIdColumnName, i)
                                .addParameter(StoredFileEntityInformation.libraryIdColumnName, 5)
                                .addParameter(StoredFileEntityInformation.isOwnerColumnName, true)
                                .execute()
                        }
                        it.setTransactionSuccessful()
                    }
                }
            }.toExpiringFuture().get()

            val storedFilesCollection =
                StoredFileAccess(ApplicationProvider.getApplicationContext())
            storedFilesCollection.promiseAllStoredFiles(LibraryId(5)).toExpiringFuture().get()
        }

		@AfterClass
		@JvmStatic
		fun cleanup() {
			storedFiles = null
		}
	}

	@Test
	fun thenTheStoredFilesAreFromTheCorrectLibrary() {
		Assertions.assertThat(storedFiles?.`object`?.map { it.libraryId }).containsOnly(5)
	}

	@Test
	fun thenTheStoredFilesAreCorrect() {
		Assertions.assertThat(storedFiles?.`object`?.map { it.serviceId.toInt() }).isSubsetOf(13..23)
	}
}
